design/cluster-api-provider-metal3/node_reuse.md

```diff
@@ -0,0 +1,278 @@
+  is WIP right now.
+
+2. be able reuse the same pool of `BareMetalHost` so that we get the storage
+  data back. Currently, there is no mechanism available in Metal³ to pick the
+  same pool of hosts that were released while upgrading/remediation - for the
+  next provisioning phase. And this proposal tries to solve it.
+
+**NOTE:** This proposal focuses on upgrade use cases but it isn't limited to
+only upgrade. Other use cases like remediation can also benefit from this
+feature as well.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I would like the same nodes to be used when they go for
+reprovisioning (i.e. deprovisioning -> provisioning) during the upgrade operation.
+
+#### Story 2
+
+As a cluster admin, I would like the same nodes to be used when they go for
+reprovisioning (i.e. deprovisioning -> provisioning) during the remediation
+operation.
+
+#### Story 3
+
+As a cluster admin, I would like any available (i.e. state: `Ready`) BMH to be
+selected randomly for a new CAPI Machine -> Metal3Machine(M3M) when nodeReuse is
+False.
+
+### Goals
+
+Add a logic to label, filter and pick those labelled BMH(s) when going through a
+re-provisioning cycle as part of the upgrade procedure.
+
+### Non-Goals
+
+Support node reuse for Machines which are created independently of Kubeadm
+Control Plane controller/Machine Deployment(KCP/MD). Currently we are not trying
+to support node reuse feature for the Machines not owned by higher level objects
+like KCP/MD. Because, if there is no KCP/MD that Machine could point to, then
+Cluster API Provider Metal3 (CAPM3) Machine controller will fail to first, set
+the label in the BMH(s) and second, to filter the BMH(s) based on the label.
+
+## Proposal
+
+We would like to propose an interface to mark BMH(s) that we want to reuse after
+deprovisioning and second, add a selection/filtering mechanism in CAPM3 Machine
+controller that will pick those BMH(s) with a specific label (to be exact, based
+on the label `infrastructure.cluster.x-k8s.io/node-reuse`).
+
+## Design Details
+
+In case of nodeReuse field is set to True, to make sure that a BMH(s) is not
+utilized by other M3M(s) after it has been deprovisioned, the new label
+`infrastructure.cluster.x-k8s.io/node-reuse` will store the name of KCP/MD even
+when it is in `Ready` state. During the upgrade, after KCP/MD creates new
+Machines, CAPM3 will create new M3M(s) and will filter out BMH(s) that have the
+name of KCP/MD in the `infrastructure.cluster.x-k8s.io/node-reuse` label.
+If the name of the KCP/MD matches the name set in
+`infrastructure.cluster.x-k8s.io/node-reuse` label of BMH, then CAPM3 Machine
+controller will associate that BMH to the M3M.
+
+As such, even after deprovisioning we will keep the
+`infrastructure.cluster.x-k8s.io/node-reuse` on a BMH and when M3M will be
+requesting a new BMH, CAPM3 will filter BMH(s) based on their label and will
+pick up those that match the name of KCP/MD.
+
+**Example flow:**
```

Comment:
More than "example flow" I think that below are reported the details of three specific scenarios (and maybe they could be moved in the `Implementation Details/Notes/Constraints` section?) - so that they have been also named properly. Are there any other relevant scenarios to be considered (for example, the simple creation)?

Reply:
We have moved them under `Implementation Details/Notes/Constraints` section.